### PR TITLE
FileUpload options for Server Config

### LINKED ISSUE #2137: Security for file uploads
### Issue Description

We've managed to set-up an S3 bucket without any issues and file uploads work great. 

My main concern is that anyone can upload files to my bucket without any security . 

I've searched all around and even looked through the adapter and haven't seen anywhere a security check . 

Testing this it turn out that anyone can make file uploads to my bucket as long as they know my appID
### Steps to reproduce

Setup an S3 bucket . 

Make a POST request to upload a file. 
For example : 
`curl -X POST -H "X-Parse-Application-Id: <APP_ID>" -H "Content-Type: text/plain" -T "D:\test.txt" http://localhost/parse/files/test.txt`
#### Expected Results

An unauthorized error as long as no session token( or any other authentication method ) was provided in the header. 
#### Actual Outcome

```
{
 "url":"http://localhost/parse/files/<APP_ID>/fc80b2e510c27d544b0d845ddcfc88e4_test.txt",
 "name":"fc80b2e510c27d544b0d845ddcfc88e4_test.txt"
}
```
### Environment Setup
- **Server**
  - parse-server version: [2.2.13]
  - Localhost or remote server? (AWS, Heroku, Azure, Digital Ocean, etc): [Local && Heroku]

### PR DESCRIPTION
**Pull request by dblythy**
As discussed in #6995, this PR allows the following options for parse server:

```
fileUpload: {
  enabled: true, // Is true if files can be uploaded with Parse Server. Default is true.
  enabledForAnonymousUser: false, // Is true if file upload is enabled for Anonymous Users. Default is false.
  enabledForPublic: false // If true if file upload is enabled for anyone with access to the Parse Server file upload endpoint, regardless of user authentication. Default is false.
}
```
(thank you @mtrezza for the copy + paste)

This is a **breaking** change as by default, public file uploads will be rejected.

Properly closes #2137, and in relation to our [discussion on the community forum.](https://community.parseplatform.org/t/parse-security-how-to-improve-default-security/924/23)

### CODE REVIEW

File: src/Options/Definitions.js [hunk 1]
```diff
@@ -168,6 +168,11 @@ module.exports.ParseServerOptions = {
     help: 'Adapter module for the files sub-system',
     action: parsers.moduleOrObjectParser,
   },
+  fileUpload: {
```
Comment (mtrezza) [inline src/Options/Definitions.js:171]: Did you manually add these lines to the `Definitions.js` file? This file is recreated by the nodejs `definitions` script, based on the definition that you add in `/src/Options/index.js`. This will also create these options for the CLI. Note that since we have sub-parameters here, you'll need to
- create an interface in `/src/Options/index.js` where you add these sub-parameters and
- add an entry to `getENVPrefix` in `buildConfigDefinitions.js`

**Review by mtrezza — CHANGES_REQUESTED** [review]

File: src/Options/Definitions.js (hunk 1 above)
Comment (dblythy) [reply src/Options/Definitions.js:171]: Oh, thanks for the pointers. I didn't realise that.

I will need to do that on the Security Vulnerability PR. I'll do both that tomorrow. Thanks!

Comment (mtrezza) [reply src/Options/Definitions.js:171]: You can take a look at an existing config object, for example search for the term `PARSE_SERVER_EXPERIMENTAL_IDEMPOTENCY` and you'll find hints for how it's done.

Comment (dblythy) [reply src/Options/Definitions.js:167]: Thanks mate. I didn't realise the blaring 

"This code has been generated by resources/buildConfigDefinitions.js
Do not edit manually, but update Options/index.js"

My bad. How does that look now? I'll also have to change the MFA code as well.

Comment (dblythy) [reply src/Options/Definitions.js:171]: I've also added `encryptionKey` to Options/index.js, as when I ran `definitions`, it got removed.

File: resources/buildConfigDefinitions.js [hunk 2]
```diff
@@ -55,6 +55,9 @@ function getENVPrefix(iface) {
   if (iface.id.name === 'IdempotencyOptions') {
     return 'PARSE_SERVER_EXPERIMENTAL_IDEMPOTENCY_';
   }
+  if (iface.id.name === 'FileUploadOptions') {
+    return 'PARSE_SERVER_PARSE_SERVER_FILE_UPLOAD_';
```
Comment (mtrezza) [inline resources/buildConfigDefinitions.js:59]: I think that's double.

**Review by mtrezza — CHANGES_REQUESTED** [review]

File: src/Options/Definitions.js (hunk 1 above)
Comment (mtrezza) [reply src/Options/Definitions.js:171]: We shouldn't mix these things. I would see that in a separate PR.

File: src/Options/index.js [hunk 3]
```diff
@@ -77,6 +77,9 @@ export interface ParseServerOptions {
   javascriptKey: ?string;
   /* Key for Unity and .Net SDK */
   dotNetKey: ?string;
+  /* Key for encrypting your files
```
Comment (mtrezza) [inline src/Options/index.js:80]: Please move this to a separate PR to keep this PR confined to its description.

**Review by mtrezza — CHANGES_REQUESTED** [review]

File: resources/buildConfigDefinitions.js (hunk 2 above)
Comment (dblythy) [reply resources/buildConfigDefinitions.js:59]: oops

**Comment by mtrezza** [comment]
What are the open points in this PR?

**Comment by dblythy** [comment]
I’ve got to resolve conflicts and fix failing tests, sorry.

**Comment by dblythy** [comment]
I accidentally rebased from my branch which I use on my AWS servers and i'm not overly experienced with git, so I just created a new branch with the changes. Sorry!!
